No overload variant of "__sub__" of "Series" matches argument type "Timedelta"

**Describe the bug**
After upgrading to pandas 2.0 started getting errors as described in the title

**To Reproduce**
```python
import pandas as pd

df = pd.read_csv("afile.csv", parse_dates=["foo"])
df["foo"] = df["foo"] - pd.Timedelta("1D")
```

Using mypy:

```sh
pandas-stubs-mre.py:4: error: No overload variant of "__sub__" of "Series" matches argument type "Timedelta"  [operator]
pandas-stubs-mre.py:4: note: Possible overload variants:
pandas-stubs-mre.py:4: note:     def __sub__(self, Timestamp | datetime | TimestampSeries, /) -> TimedeltaSeries
pandas-stubs-mre.py:4: note:     def __sub__(self, complex | ExtensionArray | ndarray[Any, Any] | dict[str, ndarray[Any, Any]] | list[Any] | tuple[Any, ...] | Index | Series[Any], /) -> Series[Any]
Found 1 error in 1 file (checked 1 source file)
```


**Please complete the following information:**
 - OS: Discovered on both Windows / Linux
 - python version: 3.11.0
 - version of type checker: mypy 1.4.1 (compiled: yes)
 - version of installed `pandas-stubs`: 2.0.2.230605


**Additional context**
The issue does not happen with older versions of pandas-stubs - 1.5.3.230321 issues no complaint. I haven't done a full bisect but guess this may go back to https://github.com/pandas-dev/pandas-stubs/pull/616 ?

--- a/pandas-stubs/core/series.pyi
+++ b/pandas-stubs/core/series.pyi
@@ -1534,6 +1534,16 @@ class Series(IndexOpsMixin, NDFrame, Generic[S1]):
     def __rxor__(self, other: int | np_ndarray_anyint | Series[int]) -> Series[int]: ...  # type: ignore[misc]
     @overload
     def __sub__(
+        self: Series[Timestamp],
+        other: Timedelta | TimedeltaSeries | TimedeltaIndex | np.timedelta64,
+    ) -> TimestampSeries: ...
+    @overload
+    def __sub__(
+        self: Series[Timedelta],
+        other: Timedelta | TimedeltaSeries | TimedeltaIndex | np.timedelta64,
+    ) -> TimedeltaSeries: ...
+    @overload
+    def __sub__(
         self, other: Timestamp | datetime | TimestampSeries
     ) -> TimedeltaSeries: ...
     @overload

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -197,6 +197,7 @@ reportPrivateUsage = false
 # enable optional checks
 reportMissingModuleSource = true
 useLibraryCodeForTypes = false
+defineConstant = { MYPY_CHECKING = false }
 
 [tool.codespell]
 ignore-words-list = "indext, mose, sav, ser"

--- a/tests/test_indexes.py
+++ b/tests/test_indexes.py
@@ -25,6 +25,8 @@ from tests import (
 
 if TYPE_CHECKING:
     MYPY_CHECKING: bool = True
+    # See test_sorted_and_list() where mypy and pyright do different
+    # inference on sorted(pd.Index)
     if MYPY_CHECKING:
         from typing import Any
 
@@ -728,6 +730,8 @@ def test_interval_index_tuples():
 def test_sorted_and_list() -> None:
     # GH 497
     i1 = pd.Index([3, 2, 1])
+    # mypy infers sorted(i1) as list[Any], while pyright infers sorted(i1) as
+    # list[SupportsRichComparison]
     check(
         assert_type(
             sorted(i1),

--- a/tests/test_indexes.py
+++ b/tests/test_indexes.py
@@ -25,14 +25,15 @@ from tests import (
 
 if TYPE_CHECKING:
     MYPY_CHECKING: bool = True
-    if MYPY_CHECKING:
+    if not MYPY_CHECKING:
+        from _typeshed import SupportsRichComparison
+    else:
         from typing import Any
 
         from typing_extensions import TypeAlias
 
         SupportsRichComparison: TypeAlias = Any
-    else:
-        pass
+
     from pandas.core.indexes.base import (
         _ComplexIndexType,
         _FloatIndexType,

--- a/tests/test_indexes.py
+++ b/tests/test_indexes.py
@@ -32,7 +32,7 @@ if TYPE_CHECKING:
 
         SupportsRichComparison: TypeAlias = Any
     else:
-        from _typeshed import SupportsRichComparison
+        pass
     from pandas.core.indexes.base import (
         _ComplexIndexType,
         _FloatIndexType,
@@ -731,7 +731,7 @@ def test_sorted_and_list() -> None:
     check(
         assert_type(
             sorted(i1),
-            list["SupportsRichComparison"],
+            "list[SupportsRichComparison]",
         ),
         list,
     )

--- a/tests/test_indexes.py
+++ b/tests/test_indexes.py
@@ -25,15 +25,14 @@ from tests import (
 
 if TYPE_CHECKING:
     MYPY_CHECKING: bool = True
-    if not MYPY_CHECKING:
-        from _typeshed import SupportsRichComparison
-    else:
+    if MYPY_CHECKING:
         from typing import Any
 
         from typing_extensions import TypeAlias
 
         SupportsRichComparison: TypeAlias = Any
-
+    else:
+        from _typeshed import SupportsRichComparison  # noqa: F401
     from pandas.core.indexes.base import (
         _ComplexIndexType,
         _FloatIndexType,

--- a/tests/test_indexes.py
+++ b/tests/test_indexes.py
@@ -24,6 +24,15 @@ from tests import (
 )
 
 if TYPE_CHECKING:
+    MYPY_CHECKING: bool = True
+    if MYPY_CHECKING:
+        from typing import Any
+
+        from typing_extensions import TypeAlias
+
+        SupportsRichComparison: TypeAlias = Any
+    else:
+        from _typeshed import SupportsRichComparison
     from pandas.core.indexes.base import (
         _ComplexIndexType,
         _FloatIndexType,
@@ -722,7 +731,7 @@ def test_sorted_and_list() -> None:
     check(
         assert_type(
             sorted(i1),
-            list,
+            list["SupportsRichComparison"],
         ),
         list,
     )

--- a/tests/test_pandas.py
+++ b/tests/test_pandas.py
@@ -1272,7 +1272,7 @@ def test_merge_ordered() -> None:
         )
         pd.merge_ordered(  # type: ignore[call-overload]
             ls,
-            rf,  # pyright: ignore[reportGeneralTypeIssues]
+            rf,
             left_on="left",
             right_on="b",
             left_by="left",  # pyright: ignore[reportGeneralTypeIssues]

--- a/tests/test_timefuncs.py
+++ b/tests/test_timefuncs.py
@@ -1130,7 +1130,14 @@ def test_timedelta64_and_arithmatic_operator() -> None:
     s1 = pd.Series(data=pd.date_range("1/1/2020", "2/1/2020"))
     s2 = pd.Series(data=pd.date_range("1/1/2021", "2/1/2021"))
     s3 = s2 - s1
-    # https://github.com/pandas-dev/pandas/issues/54059 needs to be fixed
+    check(assert_type(s3, "TimedeltaSeries"), pd.Series, pd.Timedelta)
+    td1 = pd.Timedelta(1, "D")
+    check(assert_type(s2 - td1, "TimestampSeries"), pd.Series, pd.Timestamp)
+    # GH 758
+    s4 = s1.astype(object)
+    check(assert_type(s4 - td1, "TimestampSeries"), pd.Series, pd.Timestamp)
+
+    # https://github.com/pandas-dev/pandas/issues/54059 says this is invalid
     if PD_LTE_20:
         td = np.timedelta64(1, "M")
         check(assert_type((s1 - td), "TimestampSeries"), pd.Series, pd.Timestamp)

